src/main/java/com/feedzai/commons/sql/abstraction/engine/AbstractDatabaseEngine.java

```diff
@@ -1103,6 +1104,50 @@ private void validateEntity(final DbEntity entity) throws DatabaseEngineExceptio
      */
     protected abstract boolean checkConnection(final Connection conn);
 
+    /**
+     * Executes a dummy query to verify if the connection is alive.
+     *
+     * @param conn The connection to test.
+     * @return True if the connection is valid, false otherwise.
+     * @since 2.8.0
+     */
+    protected abstract boolean pingConnection(final Connection conn);
+
+    /**
+     * Checks if the connection is alive with a given timeout.
+     * @implNote Engine aware implementations are recommended since there are some engines that do not handle
+     * {@link Connection#setNetworkTimeout(Executor, int)} synchronously, throwing unexpected exceptions, e.g:
+     * <ul>
+     * <li>Oracle</li>
+     * <li>MySQL</li>
+     *</ul>
+     *
+     * @param conn    The connection to test.
+     * @param timeout The timeout in seconds.
+     * @return {@code true} if the connection is valid, {@code false} otherwise.
+     * @since 2.8.0
+     */
+    protected boolean checkConnection(final Connection conn, final int timeout) {
+        final int socketTimeout;
+        try {
+            socketTimeout = conn.getNetworkTimeout();
+            try {
+                // Set the socket timeout to verify the connection.
+                conn.setNetworkTimeout(socketTimeoutExecutor, timeout * 1000);
+                return pingConnection(conn);
+            } catch (final Exception ex) {
+                logger.debug("Connection is down.", ex);
+                return false;
+            } finally {
+                // Make sure to respect it afterwards.
+                conn.setNetworkTimeout(socketTimeoutExecutor, socketTimeout);
+            }
+        } catch (final Exception ex) {
+            logger.debug("Connection is down.", ex);
```

Comment:
I would also try to close the connection here just in case `conn.close()`. In case the previous exception didn't set the timeout to the old one but the connection wasn't also closed